---
title: "Multiple Site Saturation Mutagenesis"
author: "Chris Ulpinnis & Pascal Püllmann"
date: "`r Sys.Date()`"
output:
  github_document: 
    html_preview: false
  md_document:
    variant: gfm
  word_document: default
  pdf_document: default
  html_document: default
vignette: >
  %\VignetteEngine{knitr}
  %\VignetteIndexEntry{Multiple Site Saturation Mutagenesis Example1 - Active Site}
  %\VignetteEncoding{UTF-8}

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
```

# Multiple Site Saturation Mutagenesis -- Active site

You can find the lastest version of this file at https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/MSD.md


## Experimental Workflow

### Target sequence
open reading frame YfeX in pCA24N (Chloramphenicol resistance)

### Clone into
seperate gene fragments into pAGM9121 first\ then reassemble into pAGM22082_cRed


### Genomic sequence YfeX

ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGT**GAC**CTGAGCGGCTTTGTT**GAC**GGTACGGAA**AAC**CCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTT**CGC**CAG**AGC**CTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA

### Restriction Enzyme
#### Level 0
BbsI

Recognition site: **_GAAGAC_**

#### Level 2
BsaI 

Recognition site: **_GGTCTC_**

### Envisioned Mutations
Aspartic Acid - 137\
Aspartic Acid - 143\
Asparagine - 147\
Arginine - 232\
Serine - 234\
\
Substitute for NDT

## R Workflow


````{r}
suppressWarnings(suppressMessages(library("GoldenMutagenesis")))
````
````{r}
input_sequence<-"ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"
recognition_site_bbsi<-"GAAGAC"
recognition_site_bsai<-"GGTCTC"
cuf<-"e_coli_316407.csv"


````
The domesticate function checks for internal cleavage sites. If corresponding sites are present silent mutations are introduced to destroy the recognition sites. 
The functions returns a list containing the position of the choosen amino acid residue for silent mutation.

````{r}
mutations_bbsi<-domesticate(input_sequence, recognition_site_bbsi, cuf=cuf)
mutations_bbsi
mutations_bsai<-domesticate(input_sequence, recognition_site_bsai, cuf=cuf)
mutations_bsai

````

The mutate_msd function designs the necessary set of primers for the desired mutations.\ 
\
The function has the following parameters:\
**prefix**: Additional nucleobases in 5' position of the recognition site [default: TT]\
**restriction_enzym**: Recognition site sequence of the respective restriction enzyme [default: GGTCTC]\
**codon**: The codon which should be used in the mutagenesis [default: NDT]\
**suffix**: Spacer nucleotides matching the cleavage pattern of the enzyme [default: A]\
**vector**: Four basepair overhangs complementary to the created overhangs in the acceptor vector  [default: c("AATG", "AAGC")]\
**replacements**: The desired substitutions as a vector with positions OR a list containing vetors with position (char) and type of MSD mutation (char) (see MSD3.rd for an example)\
**binding_min_length**: The minimal threshold value of the template binding sequence in amino acid residues [default: 4]\
**primer_length**: Maximal length of the binding sequence [default: 9]\
**target_temp**: Melting temperature of the binding sequence in °C [default: 60]\
**replacement_range**: Maximum distance between two randomization sites to be incoporated into a single primer in amino acid residues [default: 5]\
**fragment_min_size**: Minimal size of a generated gene fragment in base pairs [default 100]\
\
It will return an object of the class Primerset.\
The primers for multiple site saturation mutagenesis have an additional slot called "NDT". This slot contains a non-binding region in which (the) NDT site(s) is/are located. 

````{r}
#If domestication is necessary follow the workflow of the Point Mutagenesis vignette
mutations<-c(137,143,147,232,234)
primers<-msd_mutate(input_sequence, prefix="TT" ,restriction_enzyme=recognition_site_bsai, suffix="A", vector=c("AATG", "AAGC"), replacements=mutations, replacement_range=5, binding_min_length=4 ,primer_length=9, target_temp=60, fragment_min_size=60 )
primers
````
The primers are generated for direct cloning into the Level 2 vector.\
The function primer_add_level modifies the primers for individual cloning into Level 0 vectors and subsequent assembly in Level 2.\
The parameters **prefix, restriction_enzyme, suffix and vector** can be set similar to the mutate-function.
````{r}
primers_lvl0<-primer_add_level(primers,  prefix="TT", restriction_enzyme=recognition_site_bbsi, suffix="AA", vector=c("CTCA", "CTCG"))
primers_lvl0
````
Objects of the classes "Primer", "Primer MSD" and "Primerset" can have a slim textual output by using the function print_primer.
````{r}
print_primer(primers_lvl0)
````
The textual output can be printed into a file.
````{r eval=FALSE}
sink("primers.txt", append=FALSE, split=FALSE)
print_primer(primers_lvl0)
sink()
````