---
title: "Multiple Site Saturation Mutagenesis"
author: "Chris Ulpinnis & Pascal Püllmann"
date: "5 9 2018"
output:
  md_document:
    variant: markdown_github
  pdf_document: default
  word_document: default
  html_document: default
---
#Active site

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Multiple Site Saturation Mutagenesis

You can find the lastest version of this file at https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/MSD.md


## Experimental Workflow

### Target sequence
open reading frame YfeX in pCA24N (Chloramphenicol resistance)

### Clone into
seperate gene fragments into pAGM9121 first\ then reassemble into pAGM22082_cRed


### Genomic sequence YfeX

ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAG
TTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCT
TGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTT
CCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAA
ACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTG
GGTTGAAGAGCGT**GAC**CTGAGCGGCTTTGTT**GAC**GGTACGGAA**AAC**CCGGCGGGTGAAGAGACGCGT
CGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGA
AGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGG
CGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTT**CGC**CAG
**AGC**CTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAA
CATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTA
ACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA

### Restriction Enzyme
BbsI

Recognition site: **_GAAGAC_**

### Envisioned Mutations
Aspartic Acid - 137\
Aspartic Acid - 143\
Asparagine - 147\
Arginine - 232\
Serine - 234\
\
Substitute for NDT

## R Workflow

````{r include=FALSE}
source(file = "functions.R")
source(file = "msd_temp.R")
````
````{r eval=FALSE}
library("GoldenMutagenesis")
````
````{r}
input_sequence<-"ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"
recognition_site_bbsi<-"GAAGAC"
recognition_site_bsai<-"GGTCTC"

````
The domesticate function checks for internal cleavage sites. If corresponding sites are present silent mutations are introduced to destroy the recognition sites. 
The functions returns a list containing the position of the choosen amino acid residue for silent mutation.

````{r}
mutations_bbsi<-domesticate(input_sequence, recognition_site_bbsi)
mutations_bbsi
mutations_bsai<-domesticate(input_sequence, recognition_site_bsai)
mutations_bsai

````

The mutate_msd function designs the necessary set of primers for the desired mutations.\ 
\
The function has the following parameters:\
prefix: Additional nucleobases in 5' position of the recognition site [default: TT]\
restriction_enzyme: Recognition site sequence of the respective restriction enzyme [default: GGTCTC]\
suffix: Spacer nucleotides matching the cleavage pattern of the enzyme [default: A]\
vector: four basepair overhangs complementary to the created overhangs in the acceptor vector  [default: c("AATG", "AAGC")]\
replacements: the desired substitutions\
primer_min_length: The minimal threshold value of the template binding sequence [default: 4] in amino acid residues\
primer_length: Maximal length of the binding sequence [default: 9]\
target_temp: Melting temperature of the binding sequence in °C [default: 60]\
replacement_range: Maximum distance between two randomization sites to be incoporated into a single primer in amino acid residues [default: 5]\
fragment_min_size: Minimal size of a generated gene fragment in base pairs [default 100]\
\
It will return an object of the class Primerset.\
The primers for multiple site saturation mutagenesis have an additional slot called "NDT". This slot contains a non-binding region in which (the) NDT site(s) is/are located. 

````{r}
#If domestication is necessary follow the workflow of the Point Mutagenesis vignette
mutations<-c(137,143,147,232,234)
primers<-msd_mutate(input_sequence, prefix="TT" ,restriction_enzyme="GGTCTC", suffix="A", vector=c("AATG", "AAGC"), replacements=mutations, replacement_range=5, binding_min_length=4 ,primer_length=9, target_temp=60, fragment_min_size=60 )
primers
````
The primers are generated for direct cloning into the Level 2 vector.\
The function primer_add_level modifies the primers for individual cloning into Level 0 vectors and subsequent assembly in Level 2.\
The parameters prefix, restriction_enzyme, suffix and vector can be set similar to the mutate-function.
````{r}
primers_lvl0<-primer_add_level(primers,  prefix="TT", restriction_enzyme="GAAGAC", suffix="AA", vector=c("CTCA", "CTCG"))
primers_lvl0
````
Objects of the classes "Primer", "Primer MSD" and "Primerset" can have a slim textual output by using the function print_primer.
````{r}
print_primer(primers_lvl0)
````
The textual output can be printed into a file.
````{r eval=FALSE}
sink("primers.txt", append=FALSE, split=FALSE)
print_primer(primers_lvl0)
sink()
````