---
title: "Point Mutagenesis"
author: "Chris Ulpinnis & Pascal Püllmann"
date: "`r Sys.Date()`"
output:
  github_document: 
    html_preview: false
  md_document:
    variant: gfm
  pdf_document: default
  word_document: default
  html_document: default
vignette: >
  %\VignetteEngine{knitr}
  %\VignetteIndexEntry{Point Mutagenesis}
  %\VignetteEncoding{UTF-8}
---
# Point Mutagenesis

You can find the lastest version of this file at https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/Point_Mutagenesis.md

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Experimental Workflow

### Target sequence
open reading frame mCherry in pet28a (Kanamycin resistance)

### Clone into
pAGM9121


### Genomic sequence mCherry

ATGGTGAGCAAGGGCGAGGAGGATAACATGGCCATCATCAAGGAGTTCATGCGCTTCAAGGTGCACATGGAGGGCTCCG
TGAACGGCCACGAGTTCGAGATCGAGGGCGAGGGCGAGGGCCGCCCCTACGAGGGCACCCAGACCGCCAAGCTGAAGGT
GACCAAGGGTGGCCCCCTGCCCTTCGCCTGGGACATC**CTG**TCCCCTCAGTTCATGTACGGCTCCAAGGCCTACGT
GAAGCACCCCGCCGACATCCCCGACTACTTGAAGCTGTCCTTCCCCGAGGGCTTCAAGTGGGAGCGCGTGATGAACTTC
GAGGACGGCGGCGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGCGAGTTCATCTACAAGGTGAAGCTGCGCG
GCACCAACTTCCCCTCCGACGGCCCCGTAATGCAGAA**_GAAGAC_**GATGGGCTGGGAGGCCTCCTCCGAGCGGATGT
ACCCCGAGGACGGCGCCCTGAAGGGCGAGATCAAGCAGAGGCTGAAGCTGAAGGACGGCGGCCACTACGACGCTGAGGT
CAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGCGCCTACAACGTCAACATCAAGTTGGACATCACCTCC
CACAACGAGGACTACACCATCGTGGAACAGTACGAACGCGCCGAGGGCCGCCACTCCACCGGCGGCATGGACGAGCTGT
ACAAGGTCGACAAGCTTGCGGCCGCACTCGAGTGA


### Restriction Enzyme
BbsI

Recognition site: **_GAAGAC_**

### Envisioned Mutation
Leucine66 to Valine (point mutation)

**CTG** -> **GTG**

## R Workflow


````{r}
library("GoldenMutagenesis")
````
````{r}
input_sequence<-"ATGGTGAGCAAGGGCGAGGAGGATAACATGGCCATCATCAAGGAGTTCATGCGCTTCAAGGTGCACATGGAGGGCTCCGTGAACGGCCACGAGTTCGAGATCGAGGGCGAGGGCGAGGGCCGCCCCTACGAGGGCACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCTCAGTTCATGTACGGCTCCAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACTTGAAGCTGTCCTTCCCCGAGGGCTTCAAGTGGGAGCGCGTGATGAACTTCGAGGACGGCGGCGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGCGAGTTCATCTACAAGGTGAAGCTGCGCGGCACCAACTTCCCCTCCGACGGCCCCGTAATGCAGAAGAAGACGATGGGCTGGGAGGCCTCCTCCGAGCGGATGTACCCCGAGGACGGCGCCCTGAAGGGCGAGATCAAGCAGAGGCTGAAGCTGAAGGACGGCGGCCACTACGACGCTGAGGTCAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGCGCCTACAACGTCAACATCAAGTTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAACAGTACGAACGCGCCGAGGGCCGCCACTCCACCGGCGGCATGGACGAGCTGTACAAGGTCGACAAGCTTGCGGCCGCACTCGAGTGA"
recognition_site_bbsi<-"GAAGAC"
recognition_site_bsai<-"GGTCTC"
cuf<-"e_coli_316407.csv"

````
The domesticate function checks for internal cleavage sites. If corresponding sites are present silent mutations are introduced to destroy the recognition sites. 
The functions returns a list containing the position of the choosen amino acid residue for silent mutation.

````{r}
mutations_bbsi<-domesticate(input_sequence, recognition_site_bbsi, cuf)
mutations_bbsi
mutations_bsai<-domesticate(input_sequence, recognition_site_bsai, cuf)
mutations_bsai

````

The mutate_spm function designs the necessary set of primers for the desired mutations.\ 
\
The function has the following parameters:\
**prefix**: Additional nucleobases in 5' position of the recognition site [default: TT]\
**restriction_enzyme**: Recognition site sequence of the respective restriction enzyme [default: GGTCTC]\
**suffix**: Spacer nucleotides matching the cleavage pattern of the enzyme [default: A]\
**vector**: Four basepair overhangs complementary to the created overhangs in the acceptor vector  [default: c("AATG", "AAGC")]\
**replacements**: The desired substitutions\
**binding_min_length**: The minimal threshold value of the template binding sequence [default: 4]\
**binding_max_length**: Maximal length of the binding sequence [default: 9]\
**target_temp**: Melting temperature of the binding sequence in °C [default: 60]\
**cuf**: The Codon Usage Table which is being used to select the codon for an exchanged amino acid. [default: e_coli_316407.csv]\
**fragment_min_size**: The minimal size of a generated gene fragment in base pairs [default: 100]\
\
It will return an object of the class primer_set.

````{r}
mutations<-c(list(c(66, "V")), mutations_bbsi)
primers<-mutate_spm(input_sequence, prefix="TT", restriction_enzyme = recognition_site_bbsi, suffix = "AA", vector=c("CTCA", "CTCG"), replacements = mutations, binding_min_length=4 ,binding_max_length = 9, target_temp=60, fragment_min_size = 60, cuf=cuf)
primers
````
Objects of the classes "primer", "primer_msd" and "primer_set" can have a slim textual output by using the function print_primer.
````{r}
print_primer(primers)
````
The textual output can be printed into a file.
````{r eval=FALSE}
sink("primers.txt", append=FALSE, split=FALSE)
print_primer(primers)
sink()
````