---
title: "Quick Quality Control"
author: "Chris Ulpinnis & Pascal PÃ¼llmann"
date: "`r Sys.Date()`"
output:
  github_document: 
        html_preview: false
  md_document:
    variant: gfm
  pdf_document: default
  word_document: default
  html_document: default
vignette: >
  %\VignetteEngine{knitr}
  %\VignetteIndexEntry{Quick Quality Control}
  %\VignetteEncoding{UTF-8}
---
# Quick Quality Control

You can find the lastest version of this file at https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/QQC.md

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Experimental Workflow

### Target sequence
open reading frame YfeX in pCA24N (Chloramphenicol resistance)

### Clone into
seperate gene fragments into pAGM9121 first\ then reassemble into pAGM22082_cRed


### Genomic sequence YfeX

ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGT**GAC**CTGAGCGGCTTTGTT**GAC**GGTACGGAA**AAC**CCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTT**CGC**CAG**AGC**CTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA

### Restriction Enzyme
#### Level 0
BbsI

Recognition site: **_GAAGAC_**

#### Level 2
BsaI 

Recognition site: **_GGTCTC_**

### Envisioned Mutations
Aspartic Acid - 137\
Aspartic Acid - 143\
Asparagine - 147\
Arginine - 232\
Serine - 234\
\
Substitute for NDT


## R Workflow

### Sequence and Mutations

The target sequence and the envisioned mutations are used as input again.

````{r}
suppressWarnings(suppressMessages(library("GoldenMutagenesis")))
````
````{r}
input_sequence<-"ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"

mutations<-c(137,143,147,232,234)

````

### Quality Control
The functions aligns the obtained sequencing results to the target gene sequence. It also tries to align the reverse complement of the obtained sequence. Afterwards it checks for mismatches between the sequences.
Mismatches are likely to be sucessfully mutated nucleotides. Positions regarded as mismatches are displayed as pie charts. The shown distributions are based on the signal intensities of the four nucleobases at the mismatch positions.
You can compare the pie charts with expected pattern of randomization, therefore validating the quality of the created library.

#### Forward Sequencing
````{r eval=FALSE}
abfile<-"sequences/Yfex_0activesite_for_EF01147142.ab1")
base_distribution(input_sequence=input_sequence, ab1file=abfile, replacements=mutations)
````

````{r, echo=F}
abfile<-system.file("sequences", "Yfex_0activesite_for_EF01147142.ab1", package="GoldenMutagenesis")
base_distribution(input_sequence=input_sequence, ab1file=abfile, replacements=mutations)
````

#### Reverse Sequencing
````{r eval=FALSE}
abfile<-"sequences/Yfex_activesite_rev_EF01147143.ab1")
base_distribution(input_sequence=input_sequence, ab1file=abfile, replacements=mutations)
````

````{r, echo=F}
abfile<-system.file("sequences", "Yfex_activesite_rev_EF01147143.ab1", package="GoldenMutagenesis")
base_distribution(input_sequence=input_sequence, ab1file=abfile, replacements=mutations)
````

