<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multiple Site Saturation Mutagenesis • GoldenMutagenesis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multiple Site Saturation Mutagenesis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GoldenMutagenesis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MSD.html">Multiple Site Saturation Mutagenesis</a>
    </li>
    <li>
      <a href="../articles/MSD2.html">Multiple Site Saturation Mutagenesis Example 2</a>
    </li>
    <li>
      <a href="../articles/MSD3.html">Multiple Site Saturation Mutagenesis Example 3</a>
    </li>
    <li>
      <a href="../articles/Point_Mutagenesis.html">Point Mutagenesis</a>
    </li>
    <li>
      <a href="../articles/QQC.html">Quick Quality Control</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Multiple Site Saturation Mutagenesis</h1>
                        <h4 class="author">Chris Ulpinnis &amp; Pascal Püllmann</h4>
            
            <h4 class="date">2019-05-13</h4>
      
      
      <div class="hidden name"><code>MSD.Rmd</code></div>

    </div>

    
    
<div id="multiple-site-saturation-mutagenesis-active-site" class="section level1">
<h1 class="hasAnchor">
<a href="#multiple-site-saturation-mutagenesis-active-site" class="anchor"></a>Multiple Site Saturation Mutagenesis – Active site</h1>
<p>You can find the lastest version of this file at <a href="https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/MSD.md" class="uri">https://github.com/ipb-halle/GoldenMutagenesis/blob/master/vignettes/MSD.md</a></p>
<div id="experimental-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#experimental-workflow" class="anchor"></a>Experimental Workflow</h2>
<div id="target-sequence" class="section level3">
<h3 class="hasAnchor">
<a href="#target-sequence" class="anchor"></a>Target sequence</h3>
<p>open reading frame YfeX in pCA24N (Chloramphenicol resistance)</p>
</div>
<div id="clone-into" class="section level3">
<h3 class="hasAnchor">
<a href="#clone-into" class="anchor"></a>Clone into</h3>
<p>seperate gene fragments into pAGM9121 first then reassemble into pAGM22082_cRed</p>
</div>
<div id="genomic-sequence-yfex" class="section level3">
<h3 class="hasAnchor">
<a href="#genomic-sequence-yfex" class="anchor"></a>Genomic sequence YfeX</h3>
<p>ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGT<strong>GAC</strong>CTGAGCGGCTTTGTT<strong>GAC</strong>GGTACGGAA<strong>AAC</strong>CCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTT<strong>CGC</strong>CAG<strong>AGC</strong>CTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA</p>
</div>
<div id="restriction-enzyme" class="section level3">
<h3 class="hasAnchor">
<a href="#restriction-enzyme" class="anchor"></a>Restriction Enzyme</h3>
<div id="level-0" class="section level4">
<h4 class="hasAnchor">
<a href="#level-0" class="anchor"></a>Level 0</h4>
<p>BbsI</p>
<p>Recognition site: <strong><em>GAAGAC</em></strong></p>
</div>
<div id="level-2" class="section level4">
<h4 class="hasAnchor">
<a href="#level-2" class="anchor"></a>Level 2</h4>
<p>BsaI</p>
<p>Recognition site: <strong><em>GGTCTC</em></strong></p>
</div>
</div>
<div id="envisioned-mutations" class="section level3">
<h3 class="hasAnchor">
<a href="#envisioned-mutations" class="anchor"></a>Envisioned Mutations</h3>
<p>Aspartic Acid - 137<br>
Aspartic Acid - 143<br>
Asparagine - 147<br>
Arginine - 232<br>
Serine - 234<br><br>
Substitute for NDT</p>
</div>
</div>
<div id="r-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#r-workflow" class="anchor"></a>R Workflow</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"GoldenMutagenesis"</span>)</a></code></pre></div>
<pre><code>## Loading required package: seqinr</code></pre>
<pre><code>## Loading required package: stringr</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">input_sequence &lt;-<span class="st"> "ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"</span></a>
<a class="sourceLine" id="cb4-2" title="2">recognition_site_bbsi &lt;-<span class="st"> "GAAGAC"</span></a>
<a class="sourceLine" id="cb4-3" title="3">recognition_site_bsai &lt;-<span class="st"> "GGTCTC"</span></a>
<a class="sourceLine" id="cb4-4" title="4">cuf &lt;-<span class="st"> "e_coli_316407.csv"</span></a></code></pre></div>
<p>The domesticate function checks for internal cleavage sites. If corresponding sites are present silent mutations are introduced to destroy the recognition sites. The functions returns a list containing the position of the choosen amino acid residue for silent mutation.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">mutations_bbsi &lt;-<span class="st"> </span><span class="kw"><a href="../reference/domesticate.html">domesticate</a></span>(input_sequence, recognition_site_bbsi, </a>
<a class="sourceLine" id="cb5-2" title="2">    <span class="dt">cuf =</span> cuf)</a></code></pre></div>
<pre><code>## [1] "No domestication needed."</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">mutations_bbsi</a></code></pre></div>
<pre><code>## list()</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">mutations_bsai &lt;-<span class="st"> </span><span class="kw"><a href="../reference/domesticate.html">domesticate</a></span>(input_sequence, recognition_site_bsai, </a>
<a class="sourceLine" id="cb9-2" title="2">    <span class="dt">cuf =</span> cuf)</a></code></pre></div>
<pre><code>## [1] "No domestication needed."</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">mutations_bsai</a></code></pre></div>
<pre><code>## list()</code></pre>
<p>The mutate_msd function designs the necessary set of primers for the desired mutations. <br>
The function has the following parameters:<br><strong>prefix</strong>: Additional nucleobases in 5’ position of the recognition site [default: TT]<br><strong>restriction_enzym</strong>: Recognition site sequence of the respective restriction enzyme [default: GGTCTC]<br><strong>codon</strong>: The codon which should be used in the mutagenesis [default: NDT]<br><strong>suffix</strong>: Spacer nucleotides matching the cleavage pattern of the enzyme [default: A]<br><strong>vector</strong>: Four basepair overhangs complementary to the created overhangs in the acceptor vector [default: c(“AATG”, “AAGC”)]<br><strong>replacements</strong>: The desired substitutions as a vector with positions OR a list containing vetors with position (char) and type of MSD mutation (char) (see MSD3.rd for an example)<br><strong>binding_min_length</strong>: The minimal threshold value of the template binding sequence in amino acid residues [default: 4]<br><strong>binding_max_length</strong>: Maximal length of the binding sequence [default: 9]<br><strong>target_temp</strong>: Melting temperature of the binding sequence in °C [default: 60]<br><strong>replacement_range</strong>: Maximum distance between two randomization sites to be incoporated into a single primer in amino acid residues [default: 5]<br><strong>fragment_min_size</strong>: Minimal size of a generated gene fragment in base pairs [default 100]<br><br>
It will return an object of the class Primerset.<br>
The primers for multiple site saturation mutagenesis have an additional slot called “NDT”. This slot contains a non-binding region in which (the) NDT site(s) is/are located.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># If domestication is necessary follow the workflow of the</span></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="co"># Point Mutagenesis vignette</span></a>
<a class="sourceLine" id="cb13-3" title="3">mutations &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">137</span>, <span class="dv">143</span>, <span class="dv">147</span>, <span class="dv">232</span>, <span class="dv">234</span>)</a>
<a class="sourceLine" id="cb13-4" title="4">primers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mutate_msd.html">mutate_msd</a></span>(input_sequence, <span class="dt">prefix =</span> <span class="st">"TT"</span>, <span class="dt">restriction_enzyme =</span> recognition_site_bsai, </a>
<a class="sourceLine" id="cb13-5" title="5">    <span class="dt">suffix =</span> <span class="st">"A"</span>, <span class="dt">vector =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AATG"</span>, <span class="st">"AAGC"</span>), <span class="dt">replacements =</span> mutations, </a>
<a class="sourceLine" id="cb13-6" title="6">    <span class="dt">replacement_range =</span> <span class="dv">5</span>, <span class="dt">binding_min_length =</span> <span class="dv">4</span>, <span class="dt">binding_max_length =</span> <span class="dv">9</span>, </a>
<a class="sourceLine" id="cb13-7" title="7">    <span class="dt">target_temp =</span> <span class="dv">60</span>, <span class="dt">fragment_min_size =</span> <span class="dv">60</span>)</a>
<a class="sourceLine" id="cb13-8" title="8">primers</a></code></pre></div>
<pre><code>## An object of class "Extended_Primerset"
## Slot "fragments":
## [[1]]
## An object of class "Fragment"
## Slot "start":
## [1] 2
## 
## Slot "stop":
## [1] 142
## 
## Slot "start_mutation":
## logical(0)
## 
## Slot "stop_mutation":
## [1] 137
## 
## 
## [[2]]
## An object of class "Fragment"
## Slot "start":
## [1] 143
## 
## Slot "stop":
## [1] 231
## 
## Slot "start_mutation":
## [1] 143 147
## 
## Slot "stop_mutation":
## logical(0)
## 
## 
## [[3]]
## An object of class "Fragment"
## Slot "start":
## [1] 232
## 
## Slot "stop":
## [1] 300
## 
## Slot "start_mutation":
## [1] 232 234
## 
## Slot "stop_mutation":
## logical(0)
## 
## 
## 
## Slot "oldsequence":
## [1] "ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"
## 
## Slot "primers":
## [[1]]
## [[1]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "A"
## 
## Slot "vector":
## [1] "AATG"
## 
## Slot "overhang":
## [1] ""
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "TCTCAGGTTCAGAGTGGCATTTTGCC"
## 
## Slot "temperature":
## [1] 60.36446
## 
## Slot "difference":
## [1] 0.3644563
## 
## 
## [[1]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "T"
## 
## Slot "vector":
## [1] ""
## 
## Slot "overhang":
## [1] "ACAA"
## 
## Slot "extra":
## [1] "AGCCGCTCAGAHN"
## 
## Slot "binding_sequence":
## [1] "ACGCTCTTCAACCCAACGGAAGC"
## 
## Slot "temperature":
## [1] 59.785
## 
## Slot "difference":
## [1] 0.5794564
## 
## 
## 
## [[2]]
## [[2]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "A"
## 
## Slot "vector":
## [1] ""
## 
## Slot "overhang":
## [1] "TTGT"
## 
## Slot "extra":
## [1] "TNDTGGTACGGAANDT"
## 
## Slot "binding_sequence":
## [1] "CCGGCGGGTGAAGAGACGC"
## 
## Slot "temperature":
## [1] 59.2169
## 
## Slot "difference":
## [1] 0.7831005
## 
## 
## [[2]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "T"
## 
## Slot "vector":
## [1] ""
## 
## Slot "overhang":
## [1] "AACA"
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "ATCTTCAGCCCTTTGCCATCTTCTTT"
## 
## Slot "temperature":
## [1] 59.04605
## 
## Slot "difference":
## [1] 0.1708527
## 
## 
## 
## [[3]]
## [[3]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "A"
## 
## Slot "vector":
## [1] ""
## 
## Slot "overhang":
## [1] "TGTT"
## 
## Slot "extra":
## [1] "NDTCAGNDT"
## 
## Slot "binding_sequence":
## [1] "CTGCCGTACGGCACTGCCAG"
## 
## Slot "temperature":
## [1] 60.52991
## 
## Slot "difference":
## [1] 0.5299114
## 
## 
## [[3]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GGTCTC"
## 
## Slot "suffix":
## [1] "T"
## 
## Slot "vector":
## [1] "AAGC"
## 
## Slot "overhang":
## [1] ""
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "TTACAGCGCCATCAACTTGTCCAGC"
## 
## Slot "temperature":
## [1] 60.90327
## 
## Slot "difference":
## [1] 0.3733614
## 
## 
## 
## 
## Slot "newsequence":
## [1] "ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTNDTCTGAGCGGCTTTGTTNDTGGTACGGAANDTCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTNDTCAGNDTCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"</code></pre>
<p>The primers are generated for direct cloning into the Level 2 vector.<br>
The function primer_add_level modifies the primers for individual cloning into Level 0 vectors and subsequent assembly in Level 2.<br>
The parameters <strong>prefix, restriction_enzyme, suffix and vector</strong> can be set similar to the mutate-function.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">primers_lvl0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/primer_add_level.html">primer_add_level</a></span>(primers, <span class="dt">prefix =</span> <span class="st">"TT"</span>, <span class="dt">restriction_enzyme =</span> recognition_site_bbsi, </a>
<a class="sourceLine" id="cb15-2" title="2">    <span class="dt">suffix =</span> <span class="st">"AA"</span>, <span class="dt">vector =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"CTCA"</span>, <span class="st">"CTCG"</span>))</a>
<a class="sourceLine" id="cb15-3" title="3">primers_lvl0</a></code></pre></div>
<pre><code>## An object of class "Extended_Primerset"
## Slot "fragments":
## [[1]]
## An object of class "Fragment"
## Slot "start":
## [1] 2
## 
## Slot "stop":
## [1] 142
## 
## Slot "start_mutation":
## logical(0)
## 
## Slot "stop_mutation":
## [1] 137
## 
## 
## [[2]]
## An object of class "Fragment"
## Slot "start":
## [1] 143
## 
## Slot "stop":
## [1] 231
## 
## Slot "start_mutation":
## [1] 143 147
## 
## Slot "stop_mutation":
## logical(0)
## 
## 
## [[3]]
## An object of class "Fragment"
## Slot "start":
## [1] 232
## 
## Slot "stop":
## [1] 300
## 
## Slot "start_mutation":
## [1] 232 234
## 
## Slot "stop_mutation":
## logical(0)
## 
## 
## 
## Slot "oldsequence":
## [1] "ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"
## 
## Slot "primers":
## [[1]]
## [[1]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCA"
## 
## Slot "overhang":
## [1] "AATG"
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "TCTCAGGTTCAGAGTGGCATTTTGCC"
## 
## Slot "temperature":
## [1] 60.36446
## 
## Slot "difference":
## [1] 0.3644563
## 
## 
## [[1]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCG"
## 
## Slot "overhang":
## [1] "ACAA"
## 
## Slot "extra":
## [1] "AGCCGCTCAGAHN"
## 
## Slot "binding_sequence":
## [1] "ACGCTCTTCAACCCAACGGAAGC"
## 
## Slot "temperature":
## [1] 59.785
## 
## Slot "difference":
## [1] 0.5794564
## 
## 
## 
## [[2]]
## [[2]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCA"
## 
## Slot "overhang":
## [1] "TTGT"
## 
## Slot "extra":
## [1] "TNDTGGTACGGAANDT"
## 
## Slot "binding_sequence":
## [1] "CCGGCGGGTGAAGAGACGC"
## 
## Slot "temperature":
## [1] 59.2169
## 
## Slot "difference":
## [1] 0.7831005
## 
## 
## [[2]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCG"
## 
## Slot "overhang":
## [1] "AACA"
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "ATCTTCAGCCCTTTGCCATCTTCTTT"
## 
## Slot "temperature":
## [1] 59.04605
## 
## Slot "difference":
## [1] 0.1708527
## 
## 
## 
## [[3]]
## [[3]][[1]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCA"
## 
## Slot "overhang":
## [1] "TGTT"
## 
## Slot "extra":
## [1] "NDTCAGNDT"
## 
## Slot "binding_sequence":
## [1] "CTGCCGTACGGCACTGCCAG"
## 
## Slot "temperature":
## [1] 60.52991
## 
## Slot "difference":
## [1] 0.5299114
## 
## 
## [[3]][[2]]
## An object of class "Primer_MSD"
## Slot "prefix":
## [1] "TT"
## 
## Slot "restriction_enzyme":
## [1] "GAAGAC"
## 
## Slot "suffix":
## [1] "AA"
## 
## Slot "vector":
## [1] "CTCG"
## 
## Slot "overhang":
## [1] "AAGC"
## 
## Slot "extra":
## character(0)
## 
## Slot "binding_sequence":
## [1] "TTACAGCGCCATCAACTTGTCCAGC"
## 
## Slot "temperature":
## [1] 60.90327
## 
## Slot "difference":
## [1] 0.3733614
## 
## 
## 
## 
## Slot "newsequence":
## [1] "ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTNDTCTGAGCGGCTTTGTTNDTGGTACGGAANDTCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTNDTCAGNDTCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA"</code></pre>
<p>Objects of the classes “Primer”, “Primer MSD” and “Primerset” can have a slim textual output by using the function print_primer.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="kw"><a href="../reference/print_primer-methods.html">print_primer</a></span>(primers_lvl0)</a></code></pre></div>
<pre><code>## Fragment 1
## Start 2, Stop 142, Length 141
## Forward
## TTGAAGACAACTCAAATGTCTCAGGTTCAGAGTGGCATTTTGCC
## Temperature of binding site:  60.36446  °C 
## Temperature difference:  0.3644563  K 
## Reverse
## TTGAAGACAACTCGACAAAGCCGCTCAGAHNACGCTCTTCAACCCAACGGAAGC
## Temperature of binding site:  59.785  °C 
## Temperature difference:  0.5794564  K 
## 
## Fragment 2
## Start 143, Stop 231, Length 89
## Forward
## TTGAAGACAACTCATTGTTNDTGGTACGGAANDTCCGGCGGGTGAAGAGACGC
## Temperature of binding site:  59.2169  °C 
## Temperature difference:  0.7831005  K 
## Reverse
## TTGAAGACAACTCGAACAATCTTCAGCCCTTTGCCATCTTCTTT
## Temperature of binding site:  59.04605  °C 
## Temperature difference:  0.1708527  K 
## 
## Fragment 3
## Start 232, Stop 300, Length 69
## Forward
## TTGAAGACAACTCATGTTNDTCAGNDTCTGCCGTACGGCACTGCCAG
## Temperature of binding site:  60.52991  °C 
## Temperature difference:  0.5299114  K 
## Reverse
## TTGAAGACAACTCGAAGCTTACAGCGCCATCAACTTGTCCAGC
## Temperature of binding site:  60.90327  °C 
## Temperature difference:  0.3733614  K 
## 
## Input Sequence:
##  ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTGACCTGAGCGGCTTTGTTGACGGTACGGAAAACCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTCGCCAGAGCCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA 
## 
## Modified Sequence:
##  ATGTCTCAGGTTCAGAGTGGCATTTTGCCAGAACATTGCCGCGCGGCGATTTGGATCGAAGCCAACGTGAAAGGGGAAGTTGACGCCCTGCGTGCGGCCAGTAAAACATTTGCCGACAAACTGGCAACTTTTGAAGCGAAATTCCCGGACGCGCATCTTGGTGCGGTGGTTGCCTTTGGTAACAACACCTGGCGCGCTCTGAGCGGCGGCGTTGGGGCAGAAGAGCTGAAAGATTTTCCGGGCTACGGTAAAGGCCTTGCGCCGACGACCCAGTTCGATGTGTTGATCCACATTCTTTCTCTGCGTCACGACGTAAACTTCTCTGTCGCCCAGGCGGCGATGGAAGCCTTTGGTGACTGCATTGAAGTGAAAGAAGAGATCCACGGCTTCCGTTGGGTTGAAGAGCGTNDTCTGAGCGGCTTTGTTNDTGGTACGGAANDTCCGGCGGGTGAAGAGACGCGTCGCGAAGTGGCGGTTATCAAAGACGGCGTGGATGCGGGCGGCAGCTATGTGTTTGTCCAGCGTTGGGAACACAACCTGAAGCAGCTCAACCGGATGAGCGTTCACGATCAGGAGATGGTGATCGGGCGCACCAAAGAGGCCAACGAAGAGATCGACGGCGACGAACGTCCGGAAACCTCTCACCTCACCCGCGTTGATCTGAAAGAAGATGGCAAAGGGCTGAAGATTGTTNDTCAGNDTCTGCCGTACGGCACTGCCAGTGGCACTCACGGTCTGTACTTCTGCGCCTACTGCGCGCGTCTGCATAACATTGAGCAGCAACTGCTGAGCATGTTTGGCGATACCGATGGTAAGCGTGATGCGATGTTGCGTTTCACCAAACCGGTAACCGGCGGCTATTATTTCGCACCGTCGCTGGACAAGTTGATGGCGCTGTAA</code></pre>
<p>The textual output can be printed into a file.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sink">sink</a></span>(<span class="st">"primers.txt"</span>, <span class="dt">append =</span> <span class="ot">FALSE</span>, <span class="dt">split =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb19-2" title="2"><span class="kw"><a href="../reference/print_primer-methods.html">print_primer</a></span>(primers_lvl0)</a>
<a class="sourceLine" id="cb19-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sink">sink</a></span>()</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#multiple-site-saturation-mutagenesis-active-site">Multiple Site Saturation Mutagenesis – Active site</a><ul class="nav nav-pills nav-stacked">
<li><a href="#experimental-workflow">Experimental Workflow</a></li>
      <li><a href="#r-workflow">R Workflow</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Chris Ulpinnis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
